fluent/client/client.go

```diff
@@ -151,9 +149,33 @@ func (c *Client) SendRaw(m []byte) error {
 		return errors.New("session handshake not completed")
 	}
 
-	_, err := c.Session.Connection.Write(m)
+	var (
+		chunk string
+		err   error
+	)
+
+	if c.RequireAck {
```

Comment:
It should be like this, right? or i am missing something?
```
if c.RequireAck {
                c.ackMutex.Lock()

		if chunk, err = e.Chunk(); err != nil {
			return err
		}

		defer c.ackMutex.Unlock()
	}
```

Reply:
No. There shouldn't be a need to lock chuck because no one should be trying send the same message concurrently.

The locks are to ensure that no one can call another `Write` before the `ack` is received. That way you know the `ack` you receive is for your message.

There's a huge performance penalty for that approach, but there isn't a way around it atm. I experimented with pipelining _N_ messages together and then waiting for _N_ acks. That approach should be much faster as it allows concurrent processing. However, fluent bit returns 1 ack and then terminates the connection.

I'm not sure if it's a bug or if fluent bit doesn't support pipelining. I'm going to do a little more debugging and then report it. The only workaround I see atm is to run multiple connections.

Reply:
> However, while Fluent Bit processes the messages, it only returns 1 ack and then terminates the connection.
> 
> I'm not sure if it's a bug or if Fluent Bit doesn't support pipelining. I'm going to do a little more debugging and then report it. (It might be worth checking if it works on Fluentd) The only workaround I see atm is to run multiple connections

Definitely, worth looking into it, it should support pipelining the chunks and async responses